docs/proposals/selinux.md

```diff
@@ -0,0 +1,261 @@
+factoring for sharing volumes and other SELinux-secured resources between containers because there
+is no way in SELinux to share resources between processes with different MCS labels except to
+remove MCS labels from the shared resource.  This is a big security risk: _any container_ in the
+system can work with a resource which has the same SELinux context as it and no MCS labels.  Since
+we are also not interested in isolating containers in a pod from one another, the SELinux context
+should be shared by all containers in a pod to facilitate isolation from the containers in other
+pods and sharing resources amongst all the containers of a pod.
+
+#### Volumes
+
+Kubernetes volumes can be divided into two broad categories:
+
+1.  Unshared storage:
+    1.  Volumes created by the kubelet on the host directory: empty directory, git repo, secret,
+        downward api; all volumes in this category delegate to `EmptyDir` for their underlying
+        storage
+    2.  Volumes based on network block devices: AWS EBS, iSCSI, RBD, etc
+2.  Shared storage: NFS, Glusterfs, Cephfs, etc
+
+For unshared storage, SELinux handling for most volumes can be generalized into running a `chcon` operation on the volume directory after running the volume plugin's `Setup` function.  For these
+volumes, the Kubelet can perform the `chcon` operation and keep SELinux concerns out of the volume
+plugin code.  Some volume plugins may need to use the SELinux context during a mount operation in
+certain cases.  To account for this, our design must have a way for volume plugins to state that
+a particular volume should or should not receive generic label management. 
+
+For shared storage, the picture is murkier.  Labels for existing shared storage will be managed
+outside Kubernetes and administrators will have to set the SELinux context of pods correctly.
+The problem of solving SELinux context management for new shared storage is outside the scope for
+this proposal.
+
+## Analysis
+
+The system needs to be able to:
+
+1.  Model correctly which volumes require SELinux label management
+1.  Determine the correct ownership and SELinux context of each volume in a pod if required
+1.  Relabel volumes with the correct SELinux context when required
+
+### Modeling whether a volume requires label management
+
+#### Unshared storage: volumes derived from `EmptyDir`
+
+Empty dir and volumes derived from it are the opposite of hostPath.  Since they are created by the
+system, Kubernetes must always ensure that the ownership and SELinux context (when relevant) are
+set correctly for the volume to be usable.
+
+#### Unshared storage: network block devices
+
+Volume plugins based on network block devices such as AWS EBS and RBS can be treated the same way
+as local volumes.  Since inodes are written to these block devices in the same way as local devices,
+SELinux context can be managed on the client side by the Kubelet.
```

Comment:
these block devices can also be used in multi-reader mode and when done so are truly read-only (the OS sees it as HW read-only).